L67-cpp-callback-api.md

```diff
@@ -0,0 +1,105 @@
+* LIke the synchronnous API and unlike the completion-queue-based asynchronous API, there is no need for the application to "request" new server RPCs. Server RPC context structures will be allocated and have their resources allocated as and when RPCs arrive at the server.
+
+### Reactor model
+
+The most general form of the callback API is built around a _reactor_ model. Each type of RPC has a reactor base class provided by the library. These types are:
+
+* `ClientUnaryReactor` for unary calls
+* `ClientBidiReactor` for bidi-streaming 
+* `ClientReadReactor` for server-streaming 
+* `ClientWriteReactor` for client-streaming 
+* `ServerUnaryReactor` for unary calls
+* `ServerBidiReactor` for bidi-streaming 
+* `ServerReadReactor` for client-streaming 
+* `ServerWriteReactor` for server-streaming 
+
+Client RPC invocations from a stub provide a reactor pointer as one of their arguments, and the method handler of a server RPC must return a reactor pointer.
+
+These base classes provide three types of methods:
+
+1. Operation-initiation methods: start an asynchronous activity in the RPC. These are methods provided by the class and are not virtual.
+   - `StartRead`, `StartWrite`, `StartWritesDone`: streaming operations as appropriate to the RPC variety
+   - `StartSendInitialMetadata`: for server responses where the application wants to send the metadata separately from the first response message (uncommonly used)
+   - `Finish`: for server RPCs to provide completion status. This initiates the asynchronous transmission of the status to the client.
+1. Operation-completion reaction methods: notification of completion of asynchronous RPC activity. These are all virtual methods that default to an empty function (i.e., `{}`) but may be overridden by the application's reactor definition.
+   - `OnReadDone`, `OnWriteDone`, `OnWritesDoneDone`: methods invoked by the library at the completion of an explicitly-specified streaming RPC operation. Provides a `bool ok` argument to indicate whether the completion was successful.
+   - `OnReadInitialMetadataDone`: notifying client RPCs that the server has sent initial metadata. It has a bool `ok` argument to indicate whether the completion was successful. This is automatically invoked by the library for all RPCs; it is uncommonly used however
+   - `OnCancel`: for server RPCs only, called if an RPC is canceled before it has a chance to successfully send status to the client side
+   - `OnDone`: called when all outstanding RPC operations are completed. Provides status on the client side. 
+1. RPC completion prevention methods. These are methods provided by the class and are not virtual.
+   - `AddHold`, `RemoveHold`: for client RPCs, prevents the RPC from being considered complete (ready for `OnDone`) until each AddHold is matched to a corresponding RemoveHold. These are used to perform _extra-reaction flows_, which refers to  streaming operations initiated from outside a reaction method or method handler.
```

Comment:
We actually had an alternative proposal for such but didn't pursue it under the idea that our initial users had some experience with AddHold/RemoveHold because of some internal tooling and since that could be added either later or in a wrapper.